### PR TITLE
feat(cxx_indexer): add user-provided template instantiation excludelist

### PR DESCRIPTION
**Pull request by zrlk**
Certain files, especially in the standard library (`<type_traits>`),
generate a disproportionate amount of nearly useless data from
implicit template instantiations. This PR gives users the ability
to specify that primary/partial templates defined in certain files
should not have their implicit instantiations indexed.

### CODE REVIEW

File: kythe/cxx/indexer/cxx/IndexerASTHooks.cc [hunk 1]
```diff
@@ -1049,11 +1049,72 @@ bool IndexerASTVisitor::TraverseCXXConstructorDecl(
          TraverseDeclarationNameInfo(DNI);
 }
 
+bool IndexerASTVisitor::ShouldIndex(const clang::Decl* Decl) {
+  if (TemplateInstanceExcludePathPattern == nullptr) {
+    return true;
+  }
```
Comment (jaysachs) [inline kythe/cxx/indexer/cxx/IndexerASTHooks.cc:1055]: This is a lot of code, that conceptually is simple(-ish). I think it's worth a comment here describing that we return true if the decl is an explicit instantiation or a (partial) specialization, or if the regexp does not match the containing source file path.

File: kythe/cxx/indexer/cxx/IndexerASTHooks.h [hunk 2]
```diff
@@ -115,7 +118,14 @@ class IndexerASTVisitor : public RecursiveTypeVisitor<IndexerASTVisitor> {
         MarkedSources(&Sema, &Observer),
         ShouldStopIndexing(std::move(ShouldStopIndexing)),
         UsrByteSize(UsrByteSize),
-        DataflowEdges(EDE) {}
+        DataflowEdges(EDE) {
+    if (!TIEPP.empty()) {
+      re2::RE2::Options options;
+      options.set_never_capture(true);
+      TemplateInstanceExcludePathPattern =
+          std::make_unique<re2::RE2>(std::string(TIEPP), options);
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/IndexerASTHooks.h:126]: This silently ignores syntax errors in the regexp.  I think it makes more sense to accept a `std::unique_ptr<RE2>` (or `RE2*` or `re2::Set`) and CHECK that it is `ok()`.

Also, as an nit, there's no need to copy the `absl::string_view`; RE2 has a re2::StringPiece constructor so:

`absl::make_unique<re2::RE2>({TIEPP.data(), TIEPP.size()}, options);`

The above should work both internally and externally.

File: kythe/cxx/indexer/cxx/KytheIndexerMain.cc [hunk 3]
```diff
@@ -57,6 +57,9 @@ ABSL_FLAG(bool, use_compilation_corpus_as_default, false,
           "Use the CompilationUnit VName corpus as the default.");
 ABSL_FLAG(bool, experimental_record_dataflow_edges, false,
           "Emit experimental dataflow edges.");
+ABSL_FLAG(std::string, template_instance_exclude_path_pattern, "",
+          "If nonempty, a regex that matches files to be excluded from "
+          "template instance indexing.");
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/KytheIndexerMain.cc:62]: Per my other comment, I would convert this to a RE2 as eagerly as possible.  Ideally at argument-parsing time:
```
struct RE2Flag {
  std::unique_ptr<re2::RE2> value;
};

bool AbslParseFlag(absl::string_view text, RE2Flag* value, std::string* error) {
  re2::RE2::Options options;
  options.set_never_capture(true);
  *value = {absl::make_unique<re2::RE2>({text.data(), text.size()}, options)};
  if (!value->value->ok()) {
    *error = value->value->error_text();
  }
  return value->value.ok();
}
std::string AbslUnparseFlag(const RE2Flag& value) {
  return value.value->pattern();
}
```

File: kythe/cxx/indexer/cxx/IndexerASTHooks.cc [hunk 4]
```diff
@@ -1049,11 +1049,72 @@ bool IndexerASTVisitor::TraverseCXXConstructorDecl(
...
+            from.dyn_cast<clang::VarTemplatePartialSpecializationDecl*>()) {
+      loc = partial->getBeginLoc();
+    } else if (const auto* primary = from.dyn_cast<clang::VarTemplateDecl*>()) {
+      loc = primary->getBeginLoc();
+    } else {
+      return true;
+    }
+  } else {
+    return true;
+  }
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/IndexerASTHooks.cc:1101]: from 1056 to here should be a separate, `absl::optional<clang::SourceLocation>` returning function.

File: kythe/cxx/indexer/cxx/IndexerASTHooks.cc [hunk 5]
```diff
@@ -1049,11 +1049,72 @@ bool IndexerASTVisitor::TraverseCXXConstructorDecl(
...
+  auto file = Observer.getSourceManager()->getFilename(loc);
+  if (file.data() == nullptr) {
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/IndexerASTHooks.cc:1103]: Just `empty()`? While there may technically be a semantic distinction between `empty()` and `data() == nullptr` I don't think they're relevant here (and generally checking for a specific address on `.data()` is an anti-pattern).

File: kythe/cxx/indexer/cxx/IndexerASTHooks.cc [hunk 6]
```diff
@@ -1049,11 +1049,72 @@ bool IndexerASTVisitor::TraverseCXXConstructorDecl(
...
+      loc = partial->getBeginLoc();
+    } else if (const auto* primary = from.dyn_cast<clang::VarTemplateDecl*>()) {
+      loc = primary->getBeginLoc();
+    } else {
+      return true;
+    }
+  } else {
+    return true;
+  }
+  auto file = Observer.getSourceManager()->getFilename(loc);
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/IndexerASTHooks.cc:1102]: What happens if this `loc` is in a macro somewhere?

File: kythe/cxx/indexer/cxx/IndexerASTHooks.cc [hunk 7]
```diff
@@ -1049,11 +1049,72 @@ bool IndexerASTVisitor::TraverseCXXConstructorDecl(
...
+      return true;
+    }
+    auto from = rec->getInstantiatedFrom();
+    if (const auto* partial =
+            from.dyn_cast<clang::ClassTemplatePartialSpecializationDecl*>()) {
+      loc = partial->getBeginLoc();
+    } else if (const auto* primary =
+                   from.dyn_cast<clang::ClassTemplateDecl*>()) {
+      loc = primary->getBeginLoc();
+    } else {
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/IndexerASTHooks.cc:1081]: `from` is guaranteed to be one of the above or `nullptr`. I think doing the nullptr check first is more readable.

**Review by shahms — COMMENTED** [review]
Thanks! A few comments.

File: kythe/cxx/indexer/cxx/IndexerASTHooks.cc (hunk 6 above)
Comment (zrlk) [reply kythe/cxx/indexer/cxx/IndexerASTHooks.cc:1126]: Let's use the spelling location in this case.

File: kythe/cxx/common/re2_flag.cc [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  *value = {std::make_shared<re2::RE2>(re2::StringPiece{text.data(), text.size()}, options)};
+  if (!value->value->ok()) {
+    *error = value->value->error();
+  }
+  return value->value->ok();
+}
+std::string AbslUnparseFlag(const RE2Flag& value) {
+  return value.value->pattern();
+}
+}
```
Comment (shahms) [inline kythe/cxx/common/re2_flag.cc]: nit: newline?

**Review by shahms — APPROVED** [review]

File: kythe/cxx/common/re2_flag.cc [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  options.set_never_capture(true);
+  *value = {std::make_shared<re2::RE2>(
+      re2::StringPiece{text.data(), text.size()}, options)};
+  if (!value->value->ok()) {
+    *error = value->value->error();
+  }
+  return value->value->ok();
+}
+std::string AbslUnparseFlag(const RE2Flag& value) {
+  return value.value->pattern();
```
Comment (shahms) [inline kythe/cxx/common/re2_flag.cc:37]: need to check for nullptr :-(

File: kythe/cxx/common/re2_flag.cc [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kythe/cxx/common/re2_flag.h"
+
+namespace kythe {
+bool AbslParseFlag(absl::string_view text, RE2Flag* value, std::string* error) {
+  re2::RE2::Options options;
```
Comment (shahms) [inline kythe/cxx/common/re2_flag.cc:24]: Probably want to let the empty string mean "nullptr" as well?

**Pull request merged**